### ISSUE TITLE
Optionally keep the information about the original image in LMDB

### ISSUE THREAD

**Issue by enedene** [NVIDIA/DIGITS#630]
We've recently had an issue, we experimented with image similarity and instead returning similar images from lmdb dabatase (resized, croped and filled with noise), we wanted to return the original image, but key or value in lmdb database doesn't contain such information.

Now, we could hack the create_db.py file but we don't find this a good long term solution, so an option is to completely abandon digits or continue using it if such option becomes available.

The suggestion is to optionally provide a third column to input txt files provided to digits to fill the database, where third column would be image id. So for example key would look something like this:
b'00000234_992929_4' where 992929 would be some id provided by user, and 4 the category id. The third column should be optional for easier usage for users who do not need the info about original image.

Another option is that filepath is part of a key.

If you decide to change the key naming convention, also think about changing the first number to 9 digits so datasets of more than 100M images have valid keys, currently this looks like this (line 643 in create_db.py):

```key = '%08d_%d' % (image_count + i, datum.label)```

**Comment by gheinrich** [comment]
Hi @enedene if you create your dataset with image lists and without shuffling then I think the part of the LMDB key before the underscore is set to the index of the file in your list. Would that work for you?

**Comment by enedene** [comment]
@gheinrich the problem is that for whatever reason some images don't get loaded to database, for example a corrupted image, even with script that checks each image, it might happen that it passes yours test and not the digits and the ordering gets mixed up. For example we had 3 images in 10M+ images which didn't got into LMDB and we couldn't reconstruct the connection.

**Comment by gheinrich** [comment]
Thanks for the feedback. Assuming we fix the `index:key` mapping then this would address your concern, right?

**Comment by enedene** [comment]
@gheinrich Correct.

**Comment by gheinrich** [comment]
> If you decide to change the key naming convention, also think about changing the first number to 9 digits so datasets of more than 100M images have valid keys, currently this looks like this (line 643 in create_db.py):
> `key = '%08d_%d' % (image_count + i, datum.label)`

I don't think that would be entirely necessary:
```
print "%08d" % 123456789
123456789
```

**Comment by gheinrich** [comment]
Hi @enedene can you tell us more about your pipeline? Where would you use the LMDB key? The neural network doesn't have access to the LMDB key. Why don't you use the `test_many` button to perform inference on a list of images?

**Comment by enedene** [comment]
Hi @gheinrich for example in image similarity search. First you train the model using digits, where images are put to lmdb database, for example 256x256 with half crop, half fill transformation. Then you want to obtain the vectors inside network by running the forward pass for all the images in database, each batch you save the vectors. This part is done in caffe, then you obtain the matrix of size:

```n_images_in_lmdb x dim_of_layer_youre_interested_in ```

Each row corresponds to some image in lmdb. You can use these vectors to find the similarity.

Now lets say you have a new image, you pass it through the network, extract the vector you're interested in, compare to vectors you've extracted from your training set, select the index of image which is closest by some distance metric, then you try to use this index to find the original image (not cropped or with added noise as in half crop, half fill transformation). If index ordering is not correct, you loose the connection between original image and images in lmdb and with matrix rows which correspond to these images.

**Comment by gheinrich** [comment]
After you have trained your model, why don't you use the `test_many` button to perform inference on your list of images, instead of doing it in Caffe? This way you'd be able to map images to vectors and construct your full matrix. Then when doing single image inference it would be rather straightforward to find the most similar vector in your initial list of images, right?

**Comment by enedene** [comment]
I'm not sure if you can extract vectors by using `test_many`, you get confidence scores, which is not of any importance in this use case, we're interested in getting activations from different parts of network for every image in train set and having the ability to know from which image these activations came from.

Going through all the train images, once network was trained would be much slower as you need to make all the transformations to images you did when you were filling the database (resize, half crop, half fill). So once you train your model, database becomes useless for any further usage, this is something we'd like to avoid.

**Comment by mfernezir** [comment]
@gheinrich

Hi! To reiterate, our issue is the following. We can map each image in the database with its vector representation. We can also use a list of image file names and get their vector representations along with their classification confidences if we want them. The trouble is that we cannot map images in the database with additional image info, which we can do if we create vectors or classifications directly from files. Hence, the database is useless as a storage for anything else but simple network training. We'd like to store transformed images in a database and use that instead of raw files, for multiple other options besides network training. For instance, we could be interested in detailed layer activation for specific images in the database. Naturally, we can do that directly from files, but that is a waste of an already present database that was used for training. Simple key mapping to original files would solve this issue and make the lmdb database useful for multiple purposes.

**Comment by gheinrich** [comment]
Hi again. From what you describe you are not interested in doing image classification. What you want is to create a "generic" (also named "other" in DIGITS - admittedly not a very explicit name). The way to do it in DIGITS is to:
- create a set of LMDB databases (your job - so you're very much welcome to use any key naming convention you like)
- create a new generic dataset in DIGITS by specifying the paths to your databases
- create a new generic model and train the model on the dataset you just created

Once you're happy with your model then you can perform inference on lists of images. If you use the REST API you'll be able to programmatically capture the (vector) outputs of your model and build a large matrix out of it. I think that addresses one of your requests.

You might be interested in reviewing the [examples](https://github.com/NVIDIA/DIGITS/tree/master/examples), in particular the [Siamese](https://github.com/NVIDIA/DIGITS/tree/master/examples/siamese) example, which does something somewhat similar to what you are willing to do (mapping an image to a vector).

To address your other request (being able to perform inference on a database), I have pushed https://github.com/NVIDIA/DIGITS/pull/638, which is work in progress, but you're welcome to try it.

**Comment by mfernezir** [comment]
Thanks, we'll have a look.

Just to be more precise, we are still interested in classification. At the moment, we are creating our databases with DIGITS and doing regular classification training jobs. Additionally, we'd like to use the same database for other tasks not necessarily directly with DIGTIS. For our purposes, it could still be that it's simpler to modify key naming and to keep compatibility with regular DIGITS job creation.

**Comment by gheinrich** [comment]
Hello, any feedback on #638?

**Comment by enedene** [comment]
Hi @gheinrich we haven't got to it yet, filling the lmdb database on our dataset takes a long time, we're currently considering options. This complaint is mainly that old lmdb databases become useless for further usage.
